src/Symfony/Bridge/ProxyManager/LazyProxy/PhpDumper/ProxyDumper.php

```diff
@@ -70,13 +70,14 @@ public function getProxyFactoryCode(Definition $definition, $id, $methodName = n
         $instantiation = 'return';
 
         if ($definition->isShared()) {
-            $instantiation .= sprintf(' $this->%s[\'%s\'] =', $definition->isPublic() || !method_exists(ContainerBuilder::class, 'addClassResource') ? 'services' : 'privates', $id);
+            $instantiation .= sprintf(' $container->%s[\'%s\'] =', $definition->isPublic() || !method_exists(ContainerBuilder::class, 'addClassResource') ? 'services' : 'privates', $id);
         }
 
         if (null === $methodName) {
             throw new \InvalidArgumentException(sprintf('Missing name of method to call to construct the service "%s".', $id));
         }
 
+        $factoryCall = sprintf($definition->isPublic() || !method_exists(ContainerBuilder::class, 'addClassResource') ? '$container->%s(false)' : '%s::create($container, false)', $methodName);
```

Comment:
passing the factory class instead of the method name may need to be done another way to preserve BC